File: 2022/rust/src/bin/day14.rs
use aoc2022::Solve;
use itertools::Itertools;
use std::io::{BufRead, BufReader};

struct Solution {
    cave: Vec<Vec<bool>>,
    ymax: usize,
}

impl Solution {
    fn count_units(&self, floor: bool) -> u32 {
        let mut block = self.cave.clone();
        let mut stack = vec![(500, 0)];
        let mut units = 0;
        while let Some(&(x, y)) = stack.last() {
            if let Some(&next) = [(x, y + 1), (x - 1, y + 1), (x + 1, y + 1)]
                .iter()
                .find(|p| !block[p.0][p.1])
            {
                if y < self.ymax + 1 {
                    stack.push(next);
                    continue;
                } else if !floor {
                    break;
                }
            }
            if let Some((x, y)) = stack.pop() {
                block[x][y] = true;
                units += 1;
            }
        }
        units
    }
}

impl Solve for Solution {
    type Answer1 = u32;
    type Answer2 = u32;

    fn new(r: impl std::io::Read) -> Self {
        let mut cave = vec![vec![false; 500]; 1001];
        let mut ymax = 0;
        for line in BufReader::new(r).lines().filter_map(Result::ok) {
            for (p0, p1) in line
                .split(" -> ")
                .filter_map(|xy| {
                    xy.split(',')
                        .filter_map(|s| s.parse().ok())
                        .collect_tuple::<(usize, usize)>()
                })
                .tuple_windows()
            {
                (p0.0.min(p1.0)..=p0.0.max(p1.0)).for_each(|x| cave[x][p0.1] = true);
                (p0.1.min(p1.1)..=p0.1.max(p1.1)).for_each(|y| cave[p0.0][y] = true);
                ymax = ymax.max(p0.1.max(p1.1));
            }
        }
        Self { cave, ymax }
    }
    fn part1(&self) -> Self::Answer1 {
        self.count_units(false)
    }
    fn part2(&self) -> Self::Answer2 {
        self.count_units(true)
    }
}

fn main() {
    let solution = Solution::new(std::io::stdin().lock());
    println!("Part 1: {}", solution.part1());
    println!("Part 2: {}", solution.part2());
}

#[cfg(test)]
mod tests {
    use super::*;

    fn example_input() -> &'static [u8] {
        r"
498,4 -> 498,6 -> 496,6
503,4 -> 502,4 -> 502,9 -> 494,9
"[1..]
            .as_bytes()
    }

    #[test]
    fn part1() {
        assert_eq!(24, Solution::new(example_input()).part1());
    }

    #[test]
    fn part2() {
        assert_eq!(93, Solution::new(example_input()).part2());
    }
}

File: 2022/ocaml/lib/day14.ml
open Base
open Solution

module Solution : Solve = struct
  type state = Down of (int * int) | Rest | Void
  type t = state -> int

  let parse input =
    let cave =
      let parse_line line =
        let parse_xy s =
          let x, y = String.lsplit2_exn s ~on:',' in
          (Int.of_string x, Int.of_string y)
        in
        String.split line ~on:' '
        |> List.filter ~f:(String.( <> ) "->")
        |> List.map ~f:parse_xy |> Utils.pairs
        |> List.map ~f:(fun ((x0, y0), (x1, y1)) ->
               List.cartesian_product
                 (List.range (min x0 x1) (max x0 x1) ~stop:`inclusive)
                 (List.range (min y0 y1) (max y0 y1) ~stop:`inclusive))
        |> List.concat
        |> Hash_set.of_list (module Utils.XY)
      in
      Stdio.In_channel.input_lines input
      |> List.map ~f:parse_line
      |> List.fold ~init:(Hash_set.create (module Utils.XY)) ~f:Hash_set.union
    in
    let ymax =
      Hash_set.to_list cave |> List.map ~f:snd |> List.fold ~init:0 ~f:max
    in
    fun state ->
      let block = Hash_set.copy cave in
      let rec dfs = function
        | [] -> Hash_set.length block - Hash_set.length cave
        | hd :: tl ->
            hd
            |> (fun (x, y) -> [ (x, y + 1); (x - 1, y + 1); (x + 1, y + 1) ])
            |> List.find ~f:(Hash_set.mem block |> Fn.non)
            |> (function
                 | Some (_, y) when y > ymax + 1 -> state
                 | Some p -> Down p
                 | None -> Rest)
            |> (function
                 | Down p -> p :: hd :: tl
                 | Rest ->
                     Hash_set.add block hd;
                     tl
                 | Void -> [])
            |> dfs
      in
      dfs [ (500, 0) ]

  let part1 count_units = count_units Void |> answer_of_int
  let part2 count_units = count_units Rest |> answer_of_int
end
